cmd/utils.go

```diff
@@ -970,3 +971,164 @@ func auditLogInternal(ctx context.Context, bucket, object string, opts AuditLogO
 	logger.AuditLog(ctx, nil, nil, nil)
 }
+
+type speedtestResults []SpeedtestResult
+
+func (s speedtestResults) Len() int {
+	return len(s)
+}
+
+func (s speedtestResults) Less(i, j int) bool {
+	return s[i].Endpoint < s[i].Endpoint
+}
+
+func (s speedtestResults) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// Get the max throughput and iops numbers.
+func speedTest(ctxt context.Context, w io.Writer, throughputSize, iopsSize int, concurrencyStart int, duration time.Duration, autotune bool) (madmin.SpeedTestResult, error) {
+	var result madmin.SpeedTestResult
+
+	objAPI := newObjectLayerFn()
+	if objAPI == nil {
+		return result, errServerNotInitialized
+	}
+
+	concurrency := concurrencyStart
+
+	throughputHighestGet := uint64(0)
+	throughputHighestPut := uint64(0)
+	var throughputHighestPutResults []SpeedtestResult
+	var throughputHighestGetResults []SpeedtestResult
+
+	for {
+		// Write a blank entry to prevent client from disconnecting
+		if err := json.NewEncoder(w).Encode(madmin.SpeedTestResult{}); err != nil {
+			return result, err
+		}
+		w.(http.Flusher).Flush()
+		results := globalNotificationSys.Speedtest(ctxt, throughputSize, concurrency, duration)
+		sort.Sort(speedtestResults(results))
+		totalPut := uint64(0)
+		totalGet := uint64(0)
+		for _, result := range results {
+			totalPut += result.Uploads
+			totalGet += result.Downloads
+		}
+		if totalPut < throughputHighestPut && totalGet < throughputHighestGet {
+			break
+		}
+
+		if totalPut > throughputHighestPut {
+			throughputHighestPut = totalPut
+			throughputHighestPutResults = results
+		}
+		if totalGet > throughputHighestGet {
+			throughputHighestGet = totalGet
+			throughputHighestGetResults = results
+		}
+		if !autotune {
+			break
+		}
+		// Try with a higher concurrency to see if we get better throughput
+		concurrency += 8
+	}
+
+	concurrency = concurrencyStart
+	iopsHighestPut := uint64(0)
+	iopsHighestGet := uint64(0)
+	var iopsHighestPutResults []SpeedtestResult
+	var iopsHighestGetResults []SpeedtestResult
+
+	if autotune {
+		for {
+			// Write a blank entry to prevent client from disconnecting
+			if err := json.NewEncoder(w).Encode(madmin.SpeedTestResult{}); err != nil {
+				return result, err
+			}
+			w.(http.Flusher).Flush()
+			results := globalNotificationSys.Speedtest(ctxt, iopsSize, concurrency, duration)
+			sort.Sort(speedtestResults(results))
+			totalPut := uint64(0)
+			totalGet := uint64(0)
+			for _, result := range results {
+				totalPut += result.Uploads
+				totalGet += result.Downloads
+			}
+			if totalPut < iopsHighestPut && totalGet < iopsHighestGet {
+				break
+			}
+			if totalPut > iopsHighestPut {
+				iopsHighestPut = totalPut
+				iopsHighestPutResults = results
+			}
+			if totalGet > iopsHighestGet {
+				iopsHighestGet = totalGet
+				iopsHighestGetResults = results
+			}
+			if !autotune {
+				break
+			}
+			// Try with a higher concurrency to see if we get better throughput
+			concurrency += 8
```

Comment:
This could take a long time. Wouldn't an exponential function be better?

For example add 50% more workers:

```suggestion
			concurrency += (concurrency+1) / 2
```

Same for the other.

Reply:
Downside is it can give very inaccurate results.
Better approach is to start with a thread count such that thread_count = function(NIC_SPEED) so that we reach accurate result much faster. We could fine tune the starting thread count number later based on our experience with the speedtest in the field.

Reply:
@krishnasrinivas 

Yeah, but when you get to say 200, the increments are incredibly small, and you are much more likely to just get a random result. Lets say test run with concurrency 208 gets held up by outside IO, it will spit out 200, and you never find out that 400 is actually faster because a random hiccup stopped the test. If you make more significant increments that is much less likely to happen.